pep-0573.rst

```diff
@@ -0,0 +1,502 @@
+not hold for modules that use multi-phase initialization, so
+``PyState_FindModule`` is unavailable for these modules.
+
+A faster, safer way of accessing module-level state from extension methods
+is needed.
+
+
+Immutable Exception Types
+-------------------------
+
+For isolated modules to work, any class whose methods touch module state
+must be a heap type, so that each instance of a module can have its own
+type object.  With the changes proposed in this PEP, heap type instances will
+have access to module state without global registration.  But, to create
+instances of heap types, one will need the module state in order to
+get the type object corresponding to the appropriate module.
+In short, heap types are "viral" – anything that “touches” them must itself be
+a heap type.
+
+Curently, most exception types, apart from the ones in ``builtins``, are
+heap types.  This is likely simply because there is a convenient way
+to create them: ``PyErr_NewException``. 
+Heap types generally have a mutable ``__dict__``.
+In most cases, this mutability is harmful.  For example, exception types
+from the ``sqlite`` module are mutable and shared across subinterpreters.
+This allows "smuggling" values to other subinterpreters via attributes of
+``sqlite3.Error``.
+
+Moreover, since raising exceptions is a common operation, and heap types
+will be "viral", ``PyErr_NewException`` will tend to "infect" the module
+with "heap-type-ness" – at least if the module decides play well with
+subinterpreters/isolation.
+Many modules could go without module state
+entirely if the exception classes were immutable.
+
+To solve this problem, a new function for creating immutable exception types
+is proposed.
+
+
+Background
+===========
+
+The implementation of a Python method may need access to one or more of
+the following pieces of information:
+
+   * The instance it is called on (``self``)
+   * The underlying function
+   * The class the method was defined in
+   * The corresponding module
+   * The module state
+
+In Python code, the Python-level equivalents may be retrieved as::
+
+    import sys
+
+    class Foo:
+        def meth(self):
+            instance = self
+            module_globals = globals()
+            module_object = sys.modules[__name__]
+            underlying_function = Foo.meth
+            defining_class = Foo
+
+.. note::
+
+    The defining class is not ``type(self)``, since ``type(self)`` might
+    be a subclass of ``Foo``.
+
+Implicitly, the last three of those rely on name-based lookup via the function's ``__globals__``:
+either the ``Foo`` attribute to access the defining class and Python function object, or ``__name__`` to find the module object in ``sys.modules``.
+In Python code, this is feasible, as ``__globals__`` is set appropriately when the function definition is executed, and
+even if the namespace has been manipulated to return a different object, at worst an exception will be raised.
```

Comment:
Mention `__class__` here, but point out that the same basic analysis applies: just as reference by name relies on `__globals__` being set on the function, the cell reference relies on `__closure__` being set.

Reply:
Mentioning @encukou

Reply:
I don't think it's necessary . The Background gives the normal way to get the defining class in Python, mainly to more illustratively explain what the defining class *is*. Mentioning the class closure alternative way isn't of much use, IMO.
Also, isn't the closure a CPython implementation detail? We're explaining "This is how you do it in Python, now we want to do the same safely in C". Only the latter is CPython-specific.

Reply:
Zero-arg super() and the `__class__` closure reference are part of the language spec: https://docs.python.org/3/reference/datamodel.html?highlight=__class__#creating-the-class-object

It's passing `__classcell__` to `type.__new__` via the class execution namespace that's a CPython implementation detail. (It could be interesting to try to replace that with a context variable so it doesn't need to be explicitly preserved by metaclass authors)

Reply:
Anyway, I agree it isn't essential background for this PEP, but as someone that *does* know about `__class__`, an `# or __class__` comment after the name references would be reassuring :)

Reply:
Ha! Until a minute ago I was in that group (). Thanks for the link! Will fix.